docs/book/user-guide/llmops-guide/evaluation/retrieval.md

```diff
@@ -0,0 +1,250 @@
+and then query the PostgreSQL database for the most similar vectors. We then
+check whether the URL for the document we thought must show up is actually
+present in the top `n` results.
+
+```python
+def query_similar_docs(question: str, url_ending: str) -> tuple:
+    embedded_question = get_embeddings(question)
+    db_conn = get_db_conn()
+    top_similar_docs_urls = get_topn_similar_docs(
+        embedded_question, db_conn, n=5, only_urls=True
+    )
+    urls = [url[0] for url in top_similar_docs_urls]  # Unpacking URLs from tuples
+    return (question, url_ending, urls)
+
+def test_retrieved_docs_retrieve_best_url(question_doc_pairs: list) -> float:
+    total_tests = len(question_doc_pairs)
+    failures = 0
+
+    for pair in question_doc_pairs:
+        question, url_ending, urls = query_similar_docs(
+            pair["question"], pair["url_ending"]
+        )
+        if all(url_ending not in url for url in urls):
+            logging.error(
+                f"Failed for question: {question}. Expected URL ending: {url_ending}. Got: {urls}"
+            )
+            failures += 1
+
+    logging.info(f"Total tests: {total_tests}. Failures: {failures}")
+    failure_rate = (failures / total_tests) * 100
+    return round(failure_rate, 2)
+```
+
+We include some logging so that when running the pipeline locally we can get
+some immediate feedback logged to the console.
+
+This functionality can then be packaged up into a ZenML step once we're happy it
+does what we need:
+
+```python
+@step
+def retrieval_evaluation_small() -> Annotated[float, "small_failure_rate_retrieval"]:
+    failure_rate = test_retrieved_docs_retrieve_best_url(question_doc_pairs)
+    logging.info(f"Retrieval failure rate: {failure_rate}%")
+    return failure_rate
+```
+
+We got a 20% failure rate on the first run of this test, which was a good sign
+that the retrieval component could be improved. We only had 5 test cases, so
+this was just a starting point. In reality, you'd want to keep adding more test
+cases to cover a wider range of scenarios. You'll discover these failure cases
+as you use the system more and more, so it's a good idea to keep a record of
+them and add them to your test suite.
+
+You'd also want to examine the logs to see exactly which query failed. In our
+case, checking the logs in the ZenML dashboard, we find the following:
+
+```shell
+Failed for question: How do I generate embeddings as part of a RAG pipeline when using ZenML?. Expected URL ending: user-guide/llmops-guide/rag-with-zenml/embeddings-generation. Got: ['https://docs.zenml.io/user-guide/llmops-guide/rag-with-zenml/data-ingestion', 'https://docs.zenml.io/user-guide/llmops-guide/rag-with-zenml/understanding-rag', 'https://docs.zenml.io/v/docs/user-guide/advanced-guide/data-management/handle-custom-data-types', 'https://docs.zenml.io/user-guide/llmops-guide/rag-with-zenml', 'https://docs.zenml.io/v/docs/user-guide/llmops-guide/rag-with-zenml']
```

Comment:
not sure how to achieve this but can we make it so that this is "wrapped" and I can read it without scrolling on gitbook?